test/config/utility.cc

```diff
@@ -72,25 +95,143 @@ ConfigHelper::ConfigHelper(const Network::Address::IpVersion version) {
+  virtual_host->set_domains(0, domains);
+  virtual_host->mutable_routes(0)->mutable_match()->set_prefix(prefix);
+
+  storeHttpConnectionManager(hcm_config);
+}
+
+void ConfigHelper::setBufferLimits(uint32_t upstream_buffer_limit,
+                                   uint32_t downstream_buffer_limit) {
+  RELEASE_ASSERT(!finalized_);
+  RELEASE_ASSERT(bootstrap_.mutable_static_resources()->listeners_size() == 1);
+  auto* listener = bootstrap_.mutable_static_resources()->mutable_listeners(0);
+  listener->mutable_per_connection_buffer_limit_bytes()->set_value(downstream_buffer_limit);
+
+  auto* static_resources = bootstrap_.mutable_static_resources();
+  for (int i = 0; i < bootstrap_.mutable_static_resources()->clusters_size(); ++i) {
+    auto* cluster = static_resources->mutable_clusters(i);
+    cluster->mutable_per_connection_buffer_limit_bytes()->set_value(upstream_buffer_limit);
+  }
+
+  envoy::api::v2::filter::HttpConnectionManager hcm_config;
+  loadHttpConnectionManager(hcm_config);
+  if (hcm_config.codec_type() == envoy::api::v2::filter::HttpConnectionManager::HTTP2) {
+    uint32_t size =
+        std::max(downstream_buffer_limit, Http::Http2Settings::MIN_INITIAL_STREAM_WINDOW_SIZE);
+    auto* options = hcm_config.mutable_http2_protocol_options();
+    options->mutable_initial_stream_window_size()->set_value(size);
+    storeHttpConnectionManager(hcm_config);
+  }
+}
+
+void ConfigHelper::addRoute(const std::string& domains, const std::string& prefix,
+                            const std::string& cluster,
+                            envoy::api::v2::VirtualHost::TlsRequirementType type) {
+  RELEASE_ASSERT(!finalized_);
+  envoy::api::v2::filter::HttpConnectionManager hcm_config;
+  loadHttpConnectionManager(hcm_config);
+
+  auto* virtual_host = hcm_config.mutable_route_config()->add_virtual_hosts();
+  virtual_host->add_domains(domains);
+  virtual_host->add_routes()->mutable_match()->set_prefix(prefix);
+  virtual_host->mutable_routes(0)->mutable_route()->set_cluster(cluster);
+  virtual_host->set_require_tls(type);
+
+  storeHttpConnectionManager(hcm_config);
+}
+
+void ConfigHelper::addFilter(const std::string& config) {
+  RELEASE_ASSERT(!finalized_);
+  envoy::api::v2::filter::HttpConnectionManager hcm_config;
+  loadHttpConnectionManager(hcm_config);
+
+  auto* filter_list_back = hcm_config.add_http_filters();
+  const std::string json = Json::Factory::loadFromYamlString(config)->asJsonString();
+  MessageUtil::loadFromJson(json, *filter_list_back);
+
+  // Now move it to the front.
+  for (int i = hcm_config.http_filters_size() - 1; i > 0; --i) {
+    hcm_config.mutable_http_filters()->SwapElements(i, i - 1);
+  }
+  storeHttpConnectionManager(hcm_config);
+}
+
+void ConfigHelper::setClientCodec(envoy::api::v2::filter::HttpConnectionManager::CodecType type) {
+  RELEASE_ASSERT(!finalized_);
+  envoy::api::v2::filter::HttpConnectionManager hcm_config;
+  loadHttpConnectionManager(hcm_config);
```

Comment:
In Python I would write `with httpConnectionManager() as hcm_config: (do some stuff)`. I feel you could do something similar with either RAII or passing in a lambda, but I don't feel that strongly, this is test code and it's more aesthetics than anything.